dask/array/linalg.py

```diff
@@ -328,3 +329,98 @@ def svd(a, name=None):
     dask.array.linalg.tsqr: Actual implementation with citation
     """
     return tsqr(a, name, compute_svd=True)
+
+
+def lu(a):
+    import scipy.linalg
+
+    if a.ndim != 2:
+        raise ValueError('dimension must be 2 to perform lu decomposition')
+
+    xdim, ydim = a.shape
+    if xdim != ydim:
+        raise ValueError('input must be a square matrix to perform lu decomposition')
+    if not len(set(a.chunks[0] + a.chunks[1])) == 1:
+        raise ValueError('all chunks must be a square matrix to perform lu decomposition')
+
+    vdim = len(a.chunks[0])
+    hdim = len(a.chunks[1])
+
+    token = tokenize(a)
+    name_lu = 'lu-lu-' + token
+
+    name_p = 'lu-p-' + token
+    name_l = 'lu-l-' + token
+    name_u = 'lu-u-' + token
+
+    dsk = {}
+
+    def solve_triangular_lower(x, y):
+        try:
+            return scipy.linalg.solve_triangular(x, y, lower=True)
+        except scipy.linalg.LinAlgError:
+            # catch LinAlgError
+            # singular matrix: resolution failed at diagonal
+            return np.linalg.inv(x).dot(y)
```

Comment:
Under what conditions will the `solve_triangular` function fail?  Are there alternatives to computing the full inverse?

If possible it would be nice to avoid defining functions within other functions.  These take significantly longer to serialize when doing distributed computing.